presentation/frontend/src/presentation-frontend/StreamedResponseGenerator.ts

```diff
@@ -0,0 +1,156 @@
+    };
+  }
+
+  /**
+   * Fetches the first page.
+   * This function has to be called in order to retrieve the total items count.
+   */
+  private async fetchFirstPage(): Promise<PagedResponse<TPagedResponseItem>> {
+    const start = this._props.paging?.start ?? 0;
+    const batchSize = this._props.paging?.size ?? 0;
+    return this._props.getBatch({ start, size: batchSize }, 0);
+  }
+
+  private getRemainingPages(firstPage: PagedResponse<TPagedResponseItem>): Observable<TPagedResponseItem[]> {
+    const pageStart = this._props.paging?.start ?? 0;
+    const maxParallelRequests = this._props.maxParallelRequests;
+    const pageSize = this._props.paging?.size;
+    const { total, items: firstPageItems } = firstPage;
+
+    // If there are no items, return a single empty page.
+    if (total === 0) {
+      return of([]);
+    }
+
+    // If the response is empty, something went wrong.
+    const receivedItemsLength = firstPage.items.length;
+    if (!receivedItemsLength) {
+      handleEmptyPageResult(pageStart, total);
+    }
+
+    const totalItemsToFetch = total - pageStart;
+    if (receivedItemsLength === totalItemsToFetch) {
+      return of(firstPageItems);
+    }
+
+    let itemsToFetch: number;
+    let batchSize: number;
+    if (pageSize) {
+      itemsToFetch = Math.min(totalItemsToFetch, pageSize) - receivedItemsLength;
+      batchSize = Math.min(pageSize, receivedItemsLength);
+    } else {
+      itemsToFetch = totalItemsToFetch - receivedItemsLength;
+      batchSize = receivedItemsLength;
+    }
+
+    const remainingBatches = Math.ceil(itemsToFetch / batchSize);
+
+    // Return the first page and then stream the remaining ones.
+    return concat(
+      of(firstPage.items),
+      range(1, remainingBatches).pipe(
+        mergeMap(async (idx) => {
+          const start = pageStart + idx * batchSize;
+          const size = Math.min(total - start, batchSize);
+          const page = await this._props.getBatch({ start, size }, idx);
+          if (!page.items.length) {
+            handleEmptyPageResult(start, total);
+          }
+
+          // Pass along the index, so that the items could be sorted.
+          return { idx, items: page.items };
+        }, maxParallelRequests),
+        scan(
+          // Collect the emitted pages an emit them in the correct order.
+          (acc, value) => {
+            let { lastEmitted } = acc;
+            const { accumulatedBatches } = acc;
+            const { idx } = value;
+
+            // If current batch is not in order, put it in the accumulator
+            if (idx - 1 !== lastEmitted) {
+              accumulatedBatches.insert(value);
+              return { lastEmitted, accumulatedBatches, itemsToEmit: [] };
+            }
+
+            // Collect all batches to emit in order.
+            lastEmitted = idx;
+            const batchesToEmit = [value];
+            for (const batch of accumulatedBatches) {
+              if (batch.idx - 1 === lastEmitted) {
```

Comment:
Based on the fact that `accumulatedBatches` is sorted by batch index, I think we could `break` on `else`, no?

Reply:
That's right. I was looking for a sorted collection for the exact same point 😆